enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -8787,4010 +8787,4023 @@ class AdjointGenerator
 #endif
-          } else {
-            structAttrs[pre_args.size()] =
-#if LLVM_VERSION_MAJOR >= 12
-                // TODO persist types
-                Attribute::get(call.getContext(), "enzyme_sret_v");
-            // Attribute::get(orig->getContext(), "enzyme_sret_v",
-            // gutils->getShadowType(orig->getParamAttr(ii,
-            // Attribute::StructRet).getValueAsType()));
-#else
-                Attribute::get(call.getContext(), "enzyme_sret_v");
+                      PT = PointerType::get(
+                          anti->getType()->getPointerElementType(), AS);
+#if LLVM_VERSION_MAJOR >= 15
+                    } else {
+                      PT = PointerType::get(anti->getContext(), AS);
+                    }
 #endif
-          }
-        }
-        if (Mode != DerivativeMode::ReverseModePrimal) {
-          IRBuilder<> Builder2(call.getParent());
-          getReverseBuilder(Builder2);
+                    replacement = bb.CreateAddrSpaceCast(replacement, PT);
+                    cast<Instruction>(replacement)
+                        ->setMetadata(
+                            "enzyme_backstack",
+                            MDNode::get(replacement->getContext(), {}));
+                  }
 
-          Value *darg = nullptr;
+                  gutils->replaceAWithB(cast<Instruction>(anti), replacement);
+                  gutils->erase(cast<Instruction>(anti));
+                  anti = replacement;
+                }
+              }
 
-          if (writeOnlyNoCapture && !replaceFunction &&
-              TR.query(call.getArgOperand(i))[{-1, -1}] == BaseType::Pointer) {
-            if (EnzymeZeroCache)
-              darg =
-                  ConstantPointerNull::get(cast<PointerType>(argi->getType()));
-            else
-              darg = UndefValue::get(argi->getType());
-          } else {
-            darg = gutils->invertPointerM(call.getArgOperand(i), Builder2);
+              if (Mode == DerivativeMode::ReverseModeCombined ||
+                  (Mode == DerivativeMode::ReverseModePrimal &&
+                   forwardsShadow) ||
+                  (Mode == DerivativeMode::ReverseModeGradient &&
+                   backwardsShadow) ||
+                  (Mode == DerivativeMode::ForwardModeSplit &&
+                   backwardsShadow)) {
+                if (!inLoop) {
+                  applyChainRule(
+                      bb,
+                      [&](Value *anti) {
+                        zeroKnownAllocation(bb, anti, args, funcName,
+                                            gutils->TLI, &call);
+                      },
+                      anti);
+                }
+              }
+            }
+            gutils->invertedPointers.insert(
+                std::make_pair(&call, InvertedPointerVH(gutils, anti)));
           }
-          args.push_back(lookup(darg, Builder2));
-        }
-        pre_args.push_back(
-            gutils->invertPointerM(call.getArgOperand(
```

Comment:
Can you rename this recursivelyHandleSubfunction ?

Reply:
done